daemon/config/config_test.go

```diff
@@ -153,6 +153,32 @@ func TestDaemonConfigurationMergeConflictsWithInnerStructs(t *testing.T) {
 	}
 }
 
+// Test for #40711
+func TestDaemonConfigurationMergeFlagsWitchConfigFile(t *testing.T) {
+	f, err := ioutil.TempFile("", "docker-config-")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	configFile := f.Name()
+	f.Write([]byte(`{}`))
+	f.Close()
+
+	var conf = Config{}
+	flags := pflag.NewFlagSet("test", pflag.ContinueOnError)
+	flags.Var(&conf.NetworkConfig.DefaultAddressPools, "default-address-pool", "")
+	flags.Set("default-address-pool", "base=10.123.0.0/16,size=24")
+
+	config, err := MergeDaemonConfigurations(&conf, flags, configFile)
+
+	if err != nil {
+		t.Fatalf("merge error %v", err)
+	}
+	if config.DefaultAddressPools.Value() == nil || len(config.DefaultAddressPools.Value()) == 0 {
+		t.Fatalf("expect result not empty, but got %v", config.DefaultAddressPools.Value())
+	}
```

Comment:
Probably good to check that the value is what we expected. Something like;

```suggestion
	expected := []*ipamutils.NetworkToSplit{{Base: "10.123.0.0/16", Size: 24}}
	assert.DeepEqual(t, config.DefaultAddressPools.Value(), expected)
```

Reply:
I had the branch checked out locally, so perhaps easier to copy/paste here what I wrote up quickly with my suggestions;

```go
// Test for #40711
func TestDaemonConfigurationMergeDefaultAddressPools(t *testing.T) {
	emptyConfigFile := fs.NewFile(t, "config", fs.WithContent(`{}`))
	defer emptyConfigFile.Remove()
	configFile := fs.NewFile(t, "config", fs.WithContent(`{"default-address-pools":[{"base": "10.123.0.0/16", "size": 24 }]}`))
	defer configFile.Remove()

	expected := []*ipamutils.NetworkToSplit{{Base: "10.123.0.0/16", Size: 24}}

	t.Run("empty config file", func(t *testing.T) {
		var conf = Config{}
		flags := pflag.NewFlagSet("test", pflag.ContinueOnError)
		flags.Var(&conf.NetworkConfig.DefaultAddressPools, "default-address-pool", "")
		flags.Set("default-address-pool", "base=10.123.0.0/16,size=24")

		config, err := MergeDaemonConfigurations(&conf, flags, emptyConfigFile.Path())
		assert.NilError(t, err)
		assert.DeepEqual(t, config.DefaultAddressPools.Value(), expected)
	})

	t.Run("config file", func(t *testing.T) {
		var conf = Config{}
		flags := pflag.NewFlagSet("test", pflag.ContinueOnError)
		flags.Var(&conf.NetworkConfig.DefaultAddressPools, "default-address-pool", "")

		config, err := MergeDaemonConfigurations(&conf, flags, configFile.Path())
		assert.NilError(t, err)
		assert.DeepEqual(t, config.DefaultAddressPools.Value(), expected)
	})

	t.Run("with conflicting options", func(t *testing.T) {
		var conf = Config{}
		flags := pflag.NewFlagSet("test", pflag.ContinueOnError)
		flags.Var(&conf.NetworkConfig.DefaultAddressPools, "default-address-pool", "")
		flags.Set("default-address-pool", "base=10.123.0.0/16,size=24")

		_, err := MergeDaemonConfigurations(&conf, flags, configFile.Path())
		assert.ErrorContains(t, err, "the following directives are specified both as a flag and in the configuration file")
		assert.ErrorContains(t, err, "default-address-pools")
	})
}
```


Let me know if you think that looks good (and feel free to copy to your PR)

Reply:
It looks wonderful, really appreciate your work, I will copy it. 👍